internal/app/machined/pkg/controllers/k8s/kubelet_kubeconfig.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+)
+
+// KubeletKubeconfigController renders static pod definitions and manages k8s.KubeconfigStatus.
+type KubeletKubeconfigController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *KubeletKubeconfigController) Name() string {
+	return "k8s.KubeletKubeconfigController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *KubeletKubeconfigController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        pointer.To(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *KubeletKubeconfigController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *KubeletKubeconfigController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		machineConfig, err := r.Get(ctx, resource.NewMetadata(config.NamespaceName, config.MachineConfigType, config.V1Alpha1ID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting config: %w", err)
+		}
+
+		endpoint := machineConfig.(*config.MachineConfig).Config().Cluster().Endpoint()
+
+		if err = conditions.WaitForKubeconfigReady(constants.KubeletKubeconfig).Wait(ctx); err != nil {
+			return err
+		}
+
+		if err = updateKubeconfigControlPlaneEndpoint(endpoint.String()); err != nil {
+			return err
+		}
+	}
+}
+
+// updateKubeconfigControlPlaneEndpoint replaces the `server` address in the kubeconfig with the
+// provided new one. Returns true if the config was modified.
+func updateKubeconfigControlPlaneEndpoint(newEndpoint string) error {
+	kubeconfig, err := os.OpenFile(constants.KubeletKubeconfig, os.O_RDWR, 0)
+	if err != nil {
+		return err
+	}
+
+	defer kubeconfig.Close() //nolint:errcheck
+
+	conf := make(map[string]any)
+
+	err = yaml.NewDecoder(kubeconfig).Decode(&conf)
+	if err != nil {
+		return err
+	}
+
+	clusters, ok := conf["clusters"].([]any)
+	if !ok {
+		return fmt.Errorf("couldn't read clusters from kubeconfig")
+	}
+
+	if len(clusters) == 0 {
+		return fmt.Errorf("no clusters found in kubeconfig")
+	}
+
+	cluster, ok := clusters[0].(map[string]any)
+	if !ok {
+		return fmt.Errorf("couldn't read cluster from kubeconfig")
+	}
+
+	clusterInfo, ok := cluster["cluster"].(map[string]any)
+	if !ok {
+		return fmt.Errorf("couldn't read cluster info from kubeconfig")
+	}
+
+	server, ok := clusterInfo["server"].(string)
+	if !ok {
+		return fmt.Errorf("couldn't read server from kubeconfig")
+	}
+
+	if server == newEndpoint {
+		return nil
+	}
+
+	clusterInfo["server"] = newEndpoint
+
+	err = kubeconfig.Truncate(0)
```

Comment:
I think this might not be safe to do while the kubelet is running... as a proposal, should we rather do it in another controller?

https://github.com/AlekSi/talos/blob/2a2eba4705e0fb8ba057a753028173b79dd53af5/internal/app/machined/pkg/controllers/k8s/kubelet_service.go#L181-L190

Reply:
You suggest removing this controller completely, right?

Reply:
Moved it into `KubeletServiceController` with a bit different logic.